docs/Advanced/Sync.md

```diff
@@ -0,0 +1,163 @@
+
+### `changes` objects
+
+Changes (received from `pullChanges` and sent to `pushChanges`) are represented as an object with *raw records*. Those only use raw table and column names, and raw values (strings/numbers/booleans) — the same as in [Schema](../Schema.md).
+
+Deleted objects are always only represented by their IDs.
+
+Example:
+
+```js
+{
+  projects: {
+    created: [
+      { id: 'aaaa', name: 'Foo', is_favorite: true },
+      { id: 'bbbb', name: 'Bar', is_favorite: false },
+    ],
+    updated: [
+      { id: 'ccc', name: 'Baz', is_favorite: true },
+    ],
+    deleted: ['ddd'],
+  },
+  tasks: {
+    created: [],
+    updated: [
+      { id: 'tttt', name: 'Buy eggs' },
+    ],
+    deleted: [],
+  },
+  ...
+}
+```
+
+### `pullChanges()`
+
+Arguments: `{ lastPulledAt }`:
+- `lastPulledAt` is a timestamp for the last time client pulled changes from server (or `null` if first sync)
+
+This function should fetch from the server the list of ALL changes in all collections since `lastPulledAt`:
+- records that were created on the server
+- records that were updated on the server
+- IDs of records that were deleted on the server
+
+Return a Promise resolving to an object like this:
+
+```
+{
+  changes: { ... },
+  timestamp: 100000, // Return *server's* current time
+}
+```
+
+Raw records passed must match your app [Schema](../Schema.md), and must not contain special `_status`, `_changed` fields.
+
+The timestamp returned by the server must be a value that, if passed again to `pullChanges()` as `lastPulledAt`, will return all changes that happened since this moment.
+
+### `pushChanges()`
+
+This function will be called to push local (app) changes to the server
+
+Arguments:
+
+```
+{
+  changes: { ... },
+  // the timestamp of the last successful pull (timestamp returned in pullChanges)
+  lastSyncedAt: 10000,
+}
+```
+
+**Note:** Records sent to pushChanges might contain special `_status`, `_changed` fields. You must ignore them. You must not mutate passed changes object.
+
+`pushChanges` should call the server with the changes, and apply them remotely (create/update/delete on the server records that were created/updated/deleted locally).
+
+If successful, `pushChanges` should resolve. If there's a problem, server should revert all changes, and `pushChanges` should reject.
+
+If a record that's being pushed to the server has been changed on the server AFTER the time specified by `lastSyncedAt` (which means someone modified what we're pushing between pullChanges and pushChanges), we have a conflict, and push should be aborted. (`pushChanges` should reject). The local changes will sync during next sync.
+
+### Implementation tips
+
+Synchronization is serious business! It's very easy to make mistakes that will cause data loss. If you're not experienced at this, stick to these rules and suggestions:
+
+- **Using `synchronize()`**
+  - Ensure you never call `synchronize()` while synchronization is already in progress
```

Comment:
We make checks to ensure this is _safe_ to do so, but will still be a bad UX — either `synchronize()` will throw an error because of concurrent syncs, or it won't (depending on when exactly this happens), and sync will just waste resources